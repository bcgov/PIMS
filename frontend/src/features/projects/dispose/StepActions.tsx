import * as React from 'react';
import styled from 'styled-components';
import { Button } from 'react-bootstrap';
import useStepper from './hooks/useStepper';

const StepActionsWrapper = styled.div`
  width: 100%;
  display: flex;
  flex-direction: row-reverse;
`;

interface IStepActionsProps {
  onSave: () => void;
  onNext: () => void;
  saveDisabled?: boolean;
  nextDisabled?: boolean;
}

/**
 * A component for project disposal step actions
 * @component
 * @example
 * const api = useApi();
 * const step = useStepper();
 * const save = (data) => api.post('/project/disposal', data);
 * const onSave = () => save(data);
 * const onNext = () =>  save(data).then(() => step.next());
 * const saveDisabled = false;
 * const nextDisabled = false;
 * return (
 *  <StepActions onNext={onNext} onSave={onSave} nextDisabled={nextDisabled} saveDisabled={saveDisabled}/>
 * );
 */
export const StepActions: React.FC<IStepActionsProps> = ({
  onSave,
  onNext,
  nextDisabled,
  saveDisabled,
}) => {
  const { getNextStep } = useStepper();
  const nextLabel = getNextStep() ? 'Next' : 'Submit';
  return (
    <StepActionsWrapper>
      <Button disabled={nextDisabled} style={{ marginLeft: 10 }} onClick={onNext}>
        {nextLabel}
      </Button>
      <Button disabled={saveDisabled} onClick={onSave}>
        Save
      </Button>
    </StepActionsWrapper>
  );
};
