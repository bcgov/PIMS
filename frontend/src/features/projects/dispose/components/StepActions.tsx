import * as React from 'react';
import styled from 'styled-components';
import { Button } from 'react-bootstrap';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import Claims from 'constants/claims';

const StepActionsWrapper = styled.div`
  width: 100%;
  display: flex;
  flex-direction: row-reverse;
`;

interface IStepActionsProps {
  onSave: () => void;
  onNext: () => void;
  saveDisabled?: boolean;
  nextDisabled?: boolean;
  getNextStep?: Function;
}

/**
 * A component for project disposal step actions
 * @component
 * @example
 * const api = useApi();
 * const step = useStepper();
 * const save = (data) => api.post('/project/disposal', data);
 * const onSave = () => save(data);
 * const onNext = () =>  save(data).then(() => step.next());
 * const saveDisabled = false;
 * const nextDisabled = false;
 * return (
 *  <StepActions onNext={onNext} onSave={onSave} nextDisabled={nextDisabled} saveDisabled={saveDisabled}/>
 * );
 */
export const StepActions: React.FC<IStepActionsProps> = ({
  onSave,
  onNext,
  nextDisabled,
  saveDisabled,
  getNextStep,
}) => {
  const step = getNextStep && getNextStep();
  const nextLabel = step !== undefined ? 'Next' : 'Submit';
  const { hasClaim } = useKeycloakWrapper();
  const missingDisposeMilestonePermission = !hasClaim(Claims.DISPOSE_REQUEST) && step?.isMilestone;
  return (
    <StepActionsWrapper>
      <Button
        disabled={nextDisabled || missingDisposeMilestonePermission}
        style={{ marginLeft: 10 }}
        onClick={onNext}
      >
        {nextLabel}
      </Button>
      <Button
        variant="secondary"
        disabled={saveDisabled || missingDisposeMilestonePermission}
        onClick={onSave}
      >
        Save
      </Button>
    </StepActionsWrapper>
  );
};
