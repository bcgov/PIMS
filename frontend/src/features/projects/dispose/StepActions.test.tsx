import React from 'react';
import renderer from 'react-test-renderer';
import { StepActions } from './StepActions';
import { noop } from 'lodash';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

const mockStore = configureMockStore([thunk]);
const store = mockStore({ project: {}, projectWorkflow: [] });
const history = createMemoryHistory();

const renderComponent = (nextDisabled?: boolean, saveDisabled?: boolean) => {
  return renderer.create(
    <Provider store={store}>
      <Router history={history}>
        <StepActions
          onNext={noop}
          onSave={noop}
          nextDisabled={nextDisabled}
          saveDisabled={saveDisabled}
        />
      </Router>
    </Provider>,
  );
};

describe('Approval Confirmation', () => {
  //TODO: uncommenting either of these causes npm test to crash.
  xit('Matches Snapshot', () => {
    const component = renderComponent();
    expect(component.toJSON()).toMatchSnapshot();
  });

  xit('Action buttons are disabled', () => {
    const component = renderComponent(true, true);
    const buttons = component.root.findAllByType('button');
    expect(buttons[0].props.disabled).toBeTruthy();
    expect(buttons[1].props.disabled).toBeTruthy();
  });

  it('Action buttons are not disabled', () => {
    const component = renderComponent(false, false);
    const buttons = component.root.findAllByType('button');
    expect(buttons[0].props.disabled).toBeFalsy();
    expect(buttons[1].props.disabled).toBeFalsy();
  });
});
