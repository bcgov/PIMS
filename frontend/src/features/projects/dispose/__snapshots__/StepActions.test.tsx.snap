// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Approval Confirmation Matches Snapshot 1`] = `
<div
  className="sc-AxjAm jGlSVf"
>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "marginLeft": 10,
      }
    }
    type="button"
  >
    Submit
  </button>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Save
  </button>
</div>
`;
